torchqdynamics/utils/td_tensor.py

```diff
@@ -0,0 +1,165 @@
+        if dtype is None and not is_complex:
+            dtype = rdtype(dtype)
+        if device is None:
+            device = torch.device('cpu')
+        return CallableTDTensor(x, dtype=dtype, device=device)
+
+
+class TDTensor(ABC):
+    @abstractmethod
+    def __call__(self, t: float) -> Tensor:
+        """Evaluate a `TDTensor` at a given time"""
+        pass
+
+    @abstractmethod
+    def size(self, dim: int) -> int:
+        """Size of a `TDTensor` along a given dimension."""
+        pass
+
+    @abstractmethod
+    def dim(self) -> int:
+        """Get the number of dimensions of a `TDTensor`."""
+        pass
+
+    @abstractmethod
+    def unsqueeze(self, dim: int) -> TDTensor:
+        """Unsqueeze a `TDTensor` at position `dim`."""
+        pass
+
+    @abstractmethod
+    def has_changed(self, t: float) -> bool:
+        """Checks whether the `TDTensor` has changed since the last call."""
+        pass
+
+
+class ConstantTDTensor(TDTensor):
+    def __init__(
+        self, tensor: Tensor, *, dtype: torch.dtype, device: torch.device | None
+    ):
+        self._tensor = tensor
+        self.dtype = dtype or tensor.dtype
+        self.device = device or tensor.device
+
+    def __call__(self, t: float) -> Tensor:
+        return self._tensor
+
+    def size(self, dim: int) -> int:
+        return self._tensor.size(dim)
+
+    def dim(self) -> int:
+        return self._tensor.dim()
+
+    def unsqueeze(self, dim: int) -> ConstantTDTensor:
+        return ConstantTDTensor(
+            self._tensor.unsqueeze(dim), dtype=self.dtype, device=self.device
+        )
+
+    def has_changed(self, t: float) -> bool:
+        return False
+
+
+class CallableTDTensor(TDTensor):
+    def __init__(
+        self,
+        f: callable,
+        *,
+        dtype: torch.dtype,
+        device: torch.device,
+        shape: torch.Size | None = None,
+        check_input: bool = True,
+    ):
+        self._callable = f
+        self.dtype = dtype
+        self.device = device
+        if check_input:
+            tensor = self._check_callable()
+        self._shape = shape or tensor.shape
+        self._cached_tensor = None
+        self._last_t = None
+
+    def __call__(self, t: float) -> Tensor:
+        if t != self._last_t:
+            self._cached_tensor = self._callable(t).view(self._shape)
+            self._last_t = t
+        return self._cached_tensor
+
+    def size(self, dim: int) -> int:
+        return self._shape[dim]
+
+    def dim(self) -> int:
+        return len(self._shape)
+
+    def unsqueeze(self, dim: int) -> CallableTDTensor:
+        new_shape = list(self._shape)
+        new_shape.insert(dim, 1)
+        new_shape = torch.Size(new_shape)
+        return CallableTDTensor(
+            self._callable,
+            dtype=self.dtype,
+            device=self.device,
+            shape=new_shape,
+            check_input=False,
+        )
+
+    def has_changed(self, t: float) -> bool:
+        return True
```

Comment:
```suggestion
        return t != self._last_t
```

Reply:
No, `True` was the right thing to do. But the function name was misleading however. I changed it, and also updated the test function to make sure we're doing the intended evaluations when needed.

Reply:
Could you develop? I'm not sure I follow either

Reply:
I don't think we can just replace it. Because the `t != instance._cache[func.__name__][0]` is also important for other tensors, such as `H_nh` or `M1`. If we only run that within the `has_changed` function, these will be recomputed on every call.

Reply:
With the current implementation, there has no double checks for caching and everything is handled. However, the caching of `H` and of the other `depends_on_H` is taken care of in different places. This is the only thing that could be improved IMO, but I'm not sure how or if its needed.

Reply:
So, just to make it clearer, because the current implementation is not trivial (it took me a while to fix it). There are multiple boolean checks all around, which are (in order):

1. `solver.py` > `depends_on_H` > `if t != instance._cache[func.__name__][0]`
2. `td_tensor.py` > `if TDTensor.requires_updates(t)`
3. `td_tensor.py` > `CallableTDTensor.__call__(t)` > `if t != self._cached_t`

- Check (1) is first called everytime a `depends_on_H` function is called. If False, it means the current tensor was already cached at the current time and we use the cached version. If True, we never computed this particular tensor for the current time, and we move on to check (2).
- Check (2) is always False for `ConstantTDTensor` since the Hamiltonian is time-independent. For `CallableTDTensor`, it is always True and we move on to check (3). [NB: for PWC Hamiltonians, I guess it will be sometimes True and sometimes False.]
- Check (3) checks whether we have already computed the Hamiltonian at the current time or not. This way, the Hamiltonian is never recomputed twice for any given time, even if multiple `depends_on_H` functions are called.